### DISCUSSION TITLE
Middleware dependencies

### DISCUSSION THREAD

**Discussion by RobRoseKnows** [trpc/trpc#1503] [Q&A]
I was wondering whether tRPC has a way of defining Middleware dependencies?

An example:

Suppose I have a middleware that computes a Request Id that's shared across a request and injects it into the context, and another logging middleware that has a dependency on the Request Id being in the context. Is there a way to define the middleware such that the type checking will fail when the request id middleware comes after the logging middleware?

**Comment by KATT** [comment]
Hey!

This is one of the things that are quite tricky with the current design of tRPC. I know @anthonyshort / @gabrielsch have done some stuff with a pipe library that maybe they could share?

**Reply by KATT** [reply]
See here: https://discord.com/channels/867764511159091230/894984069291581511/915289381470883870

**Reply by RobRoseKnows** [reply]
Yeah that looks interesting, and it seems like they have the same use case as me in creating reusable Middlewares.

**Reply by anthonyshort** [reply]
We ended up just writing a helper type to make it easier to set it up. We wanted to pull all of the middleware out so that we can just pick and choose what we needed per router. The goal was to then be able to share routers across a few apps. 

Let me know if you find a more elegant solution than we found.
